src/crypto/CryptStateEVP.h

```diff
@@ -0,0 +1,287 @@
+// Copyright 2021 The Mumble Developers. All rights reserved.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file at the root of the
+// Mumble source tree or at <https://www.mumble.info/LICENSE>.
+
+// This code aims to provide a unified interface to operate the EVP interface of OpenSSL.
+// See https://wiki.openssl.org/index.php/EVP_Authenticated_Encryption_and_Decryption
+
+#ifndef MUMBLE_CRYPTOSTATEEVP_H
+#define MUMBLE_CRYPTOSTATEEVP_H
+
+#include "CryptState.h"
+
+#include <openssl/evp.h>
+#include <openssl/rand.h>
+
+
+template< class T > class CryptStateEVP : public CryptState {
+public:
+	explicit CryptStateEVP(const EVP_CIPHER *c);
+	~CryptStateEVP() override;
+
+	bool isValid() const override;
+	void genKey() override;
+	bool setKey(const std::string &rkey, const std::string &eiv, const std::string &div) override;
+	bool setRawKey(const std::string &rkey) override;
+	bool setEncryptIV(const std::string &iv) override;
+	bool setDecryptIV(const std::string &iv) override;
+	std::string getRawKey() override;
+	std::string getEncryptIV() override;
+	std::string getDecryptIV() override;
+
+	bool decrypt(const unsigned char *source, unsigned char *dst, unsigned int encrypted_length,
+				 unsigned int &plain_length) override;
+	bool encrypt(const unsigned char *source, unsigned char *dst, unsigned int plain_length,
+				 unsigned int &encrypted_length) override;
+
+private:
+	static constexpr unsigned int IV_LENGTH  = 0;
+	static constexpr unsigned int KEY_LENGTH = 0;
+	static constexpr unsigned int TAG_LENGTH = 0;
+
+	const EVP_CIPHER *cipher;
+
+	unsigned char *raw_key;
+	unsigned char *encrypt_iv;
+	unsigned char *decrypt_iv;
+	unsigned char *decrypt_history;
+};
+
+
+template< typename T > CryptStateEVP< T >::CryptStateEVP(const EVP_CIPHER *c) : CryptState(), cipher(c) {
+	raw_key         = new unsigned char[KEY_LENGTH];
+	encrypt_iv      = new unsigned char[IV_LENGTH];
+	decrypt_iv      = new unsigned char[IV_LENGTH];
+	decrypt_history = new unsigned char[0x100];
```

Comment:
I explicitly allocate these members here, because a weird bug would happen if I use automatic variables like `unsigned char raw_key[KEY_LENGTH]` that causes all four arrays here to have the same address. Weird enough, really.

Reply:
Alright... This is not some weird bug. This is just because things like `KEY_LENGTH` was not defined with a meaningful value when this class is being constructed.

Reply:
Note also that variable-size arrays on the stack are not supported in Standard C++. They are supported by GCC (and Clang I think) but last time I checked they were unsupported by MSVC.

That being said though, you should consider using smart-pointers (`std::unique_ptr`) for these arrays to avoid having to do manual memory management.

Reply:
Actually however I the `*_LENGTH` are constexpr, aren't they? In that case this _should_ work (the arrays on the stack I mean). At least as long as the sizes don't get too big. What order of magnitude are we talking about here for these sizes?

Reply:
A few bytes. Not a big deal. The problem arises when different ciphers require different values of `*_LENGTH`, like the `TAG_LENGTH` for AES-156-GCM is 8 while for CHACHA it is 16.

I would like to override these values(originally declared in `CryptStateEVP`) differently from both `CryptStateAES256GCM` and `CryptStateChacha20Poly1305`, but override a `const` value is tricky because of object slicing. That's the reason I use a template for `CryptStateEVP`.

Reply:
Okay so from what I understood these lengths are always constant but depend on which cipher is used exactly. Using templates for this seems logical but I think we might be able to make it look/work a bit nicer:
```cpp
template< typename CipherConstants > class CryptStateEVP {
    // ...
protected:
    using byte = unsigned char; // Once we upgrade to Cpp17 we can use std::byte instead
    std::array< byte, CipherConstants::KEY_LENGTH > raw_key;
    std::array< byte, CipherConstants::IV_LENGTH > encrypt_iv;
    std::array< byte, CipherConstants::IV_LENGTH > decrypt_iv;
    std::array< byte, 0x100 > decrypt_history;

    // ...
};
```
And then have e.g.
```cpp
struct AES_156_GCM_CipherConstants {
    static constexpr unsigned int KEY_LENGTH = 4;
    static constexpr unsigned int IV_LENGTH  = 8;
    // ...
};
```

And then when instantiating `CryptStateEVP` for any given cipher, you can do
```cpp
CryptStateEVP< AES_156_GCM_CipherConstants >
```

The exact naming can probably be improved upon (e.g. do we really need the suffix `CipherConstants`?)  but I hope this illustrates my idea.

Reply:
Yeah, I think `CryptStateEVP< AES_156_GCM_CipherConstants >` or `CryptStateEVP< EVPCipher::AES_156_GCM >` (whose name is preferred by me) can also be a solution, but I would also like to keep these template things inside and present merely a final `CryptStateAES256GCM` to the outside world. Meanwhile, I can't rule out some EVP cipher would need special initialization procedures so I'd rather not assume everything can be encapsulated in `EVPCipher::XXXX`. 

Though I can also store these additional steps in a method inside `EVPCipher::XXXX` (if there is any) and let the template call that method but does it sound a little bit cumbersome? 😅 While if I just inherit the template, I have more freedom to override whatever method I like. 

However, I also acknowledge my current way (inherit a template while putting itself into the `<>`) looks somewhat hacky...

Reply:
The nice thing about my suggestion is that we don't have to allocate anything on the heap. All arrays can be constructed with compile-time constant sizes.

And I think we can have the best of all worlds: For now the only difference is in some constants, so let's encapsulate them in some dummy objects as outlined by me (I also prefer your naming scheme of using `EVPCipher::XXXX`) and then we can add the respective typedefs for the specific types:
```cpp
using CryptStateAES256GCM = CryptStateEVP< EVPCipher::AES_156_GCM >;
// ...
```

And once we actually need to override functions, the new cipher can then be an actual class extending `CryptStateEVP` instead of merely being a typedef. For that to work we'd only have to make sure the respective functions are declared `virtual` in the base class (unless we will never access these objects via their super-type).

Reply:
Yeah. This sounds good. We can still extend `CryptStateEVP` if needed by inheriting.

Reply:
I'm trying this idea but somehow the compiler complains about static variables being redefined for some test cases, while it does work it I don't include those test cases. I'm confused since the `#ifndef` should prevent them from being redefined. Do you have any idea about what is wrong?

Reply:
I see the problem. The definitions of these variables were compiled into multiple object files and the problems arise when linking. I have moved these definitions into a separate .cpp file.

Reply:
> AFAIK there's no way to get the underlying c-array from std::array

Yes there is:
https://en.cppreference.com/w/cpp/container/array/data

> The definitions of these variables were compiled into multiple object files

Which variables are we talking about? :eyes: